pkg/types/binary_test.go

```diff
@@ -0,0 +1,25 @@
+package types
+
+import (
+	"encoding/json"
+	"github.com/stretchr/testify/require"
+	"testing"
+)
+
+func TestBinary_MarshalJSON(t *testing.T) {
+	assertBinaryMarshalJSON(t, nil, nil)
+	assertBinaryMarshalJSON(t, make(Binary, 0, 1), nil)
+	assertBinaryMarshalJSON(t, Binary(" "), "20")
+}
+
+func assertBinaryMarshalJSON(t *testing.T, input Binary, expect any) {
+	t.Helper()
+
+	jsn, err := input.MarshalJSON()
+	require.NoError(t, err)
+
+	var actual any
+	require.NoError(t, json.Unmarshal(jsn, &actual))
+
+	require.Equal(t, expect, actual)
```

Comment:
The use of `require` here will result in the entire test case being terminated on failure. Currently, a failing test for `nil` would stop the test for `" "` from running entirely, so a sub-test per input would be nicer. Personally, I like a slice of structs with test case name, input, expected output that's looped over starting individual sub-tests.